core/src/main/java/io/grpc/CallOptions.java

```diff
@@ -197,47 +218,55 @@ public String getCompressor() {
   @Nullable
   @ExperimentalApi("https://github.com/grpc/grpc-java/issues/1767")
   public String getAuthority() {
-    return authority;
+    return getOption(AUTHORITY_KEY);
   }
 
   /**
    * Returns a new {@code CallOptions} with {@code executor} to be used instead of the default
    * executor specified with {@link ManagedChannelBuilder#executor}.
    */
   public CallOptions withExecutor(Executor executor) {
-    CallOptions newOptions = new CallOptions(this);
-    newOptions.executor = executor;
-    return newOptions;
+    return withOption(EXECUTOR_KEY, executor);
   }
 
   @Nullable
   public Executor getExecutor() {
-    return executor;
+    return getOption(EXECUTOR_KEY);
   }
 
+  /**
+   * Create a copy of this CallOptions object with key set to value.
+   */
+  public <T> CallOptions withOption(Key<T> key, T value) {
+    CallOptions newOptions = new CallOptions(this);
```

Comment:
Not sure I follow.  The contents of the hashmap are copied in the constructor and the value of this option is replaced, if already in the map.  Are you concerned about performance with copying the hashmap?